### ISSUE TITLE
segfault when running in uwsgi

### ISSUE THREAD

**Issue by yofreke** [ldx/python-iptables#175]
Hi, I am experiencing an issue while running in uwsgi.  My code works when run using `python bin.py`, but when my server is running with uwsgi it segfaults when trying to set `match.dport = "12345"`.  Was wondering is this a problem on my end or with uwsgi compatibility?

versions:
python: 2.7.10
uwsgi: 2.0.9 and 2.0.12
python-iptables: a0cafb43c1f48a49da1e30cb83e5e0c3608b1491
iptables: v1.4.21

```
!!! uWSGI process 8249 got Segmentation Fault !!!
*** backtrace of 8249 ***
uwsgi(uwsgi_backtrace+0x2e) [0x4650ee]
uwsgi(uwsgi_segfault+0x21) [0x4654b1]
/lib/x86_64-linux-gnu/libc.so.6(+0x36d40) [0x7f300dd65d40]
/lib/x86_64-linux-gnu/libc.so.6(strlen+0x2a) [0x7f300ddb7aea]
/lib/x86_64-linux-gnu/libc.so.6(__strdup+0xe) [0x7f300ddb782e]
/lib/xtables/libxt_tcp.so(+0xd8c) [0x7f30080bed8c]
/lib/xtables/libxt_tcp.so(+0x15c4) [0x7f30080bf5c4]
/home/weebyman/manager/server/debug-env/lib/python2.7/site-packages/libxtwrapper.so(wrap_parse+0x56) [0x7f300aa30c46]
/usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_call_unix64+0x4c) [0x7f300b89dadc]
/usr/lib/x86_64-linux-gnu/libffi.so.6(ffi_call+0x1fc) [0x7f300b89d40c]
/home/weebyman/manager/server/debug-env/lib/python2.7/lib-dynload/_ctypes.x86_64-linux-gnu.so(_ctypes_callproc+0x48e) [0x7f300bab476e]
/home/weebyman/manager/server/debug-env/lib/python2.7/lib-dynload/_ctypes.x86_64-linux-gnu.so(+0x1610e) [0x7f300bab610e]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0xf84) [0x7f300e261f24]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x57ae) [0x7f300e26674e]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x80d) [0x7f300e2670dd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x1c5ff0) [0x7f300e2b9ff0]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x192a) [0x7f300e2628ca]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x80d) [0x7f300e2670dd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x517a) [0x7f300e26611a]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x57ae) [0x7f300e26674e]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x57ae) [0x7f300e26674e]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x80d) [0x7f300e2670dd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x1c5ff0) [0x7f300e2b9ff0]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0xb9dfd) [0x7f300e1addfd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x9cb53) [0x7f300e190b53]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x9d5b8) [0x7f300e1915b8]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_SetAttr+0x8f) [0x7f300e23748f]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x2263) [0x7f300e263203]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x80d) [0x7f300e2670dd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x517a) [0x7f300e26611a]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x80d) [0x7f300e2670dd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x1c60c5) [0x7f300e2ba0c5]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x192a) [0x7f300e2628ca]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x80d) [0x7f300e2670dd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x1c60c5) [0x7f300e2ba0c5]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalFrameEx+0x192a) [0x7f300e2628ca]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_EvalCodeEx+0x80d) [0x7f300e2670dd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0x1c60c5) [0x7f300e2ba0c5]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(+0xb9dfd) [0x7f300e1addfd]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyObject_Call+0x43) [0x7f300e26bde3]
/usr/lib/x86_64-linux-gnu/libpython2.7.so.1.0(PyEval_CallObjectWithKeywords+0x47) [0x7f300e221637]
/home/weebyman/manager/server/debug-env/local/lib/python2.7/site-packages/greenlet.so(+0x375c) [0x7f300caf575c]
/home/weebyman/manager/server/debug-env/local/lib/python2.7/site-packages/greenlet.so(+0x30a6) [0x7f300caf50a6]
[0x7f300cdd9280]
*** end of backtrace ***
DAMN ! worker 1 (pid: 8249) died :( trying respawn ...
```

Best I can figure is:
- `libxtwrapper.so(wrap_parse+0x56)` https://github.com/ldx/python-iptables/blob/master/libxtwrapper/wrapper.c#L31
- `/lib/xtables/libxt_tcp.so(+0x15c4)` https://github.com/qris/iptables/blob/master/extensions/libxt_tcp.c#L157
- `/lib/xtables/libxt_tcp.so(+0xd8c)` https://github.com/qris/iptables/blob/master/extensions/libxt_tcp.c#L40

**Comment by yofreke** [comment]
Update:
Tried gdb, no luck. Tried lldb, no luck.  Tried lldb-3.8, money.

Determined that the issue has to do with [optarg](http://linux.die.net/man/3/optarg)

Not garbage trace:
```
* thread #1: tid = 5025, 0x00007f3b2fbe0aea libc.so.6`__GI_strlen + 42 at strlen.S:106, name = 'uwsgi', stop reason = signal SIGSEGV: invalid address (fault address: 0x0)
  * frame #0: 0x00007f3b2fbe0aea libc.so.6`__GI_strlen + 42 at strlen.S:106
    frame #1: 0x00007f3b2fbe082e libc.so.6`__GI___strdup(s=0x0000000000000000) + 14 at strdup.c:41
    frame #2: 0x00007f3b2a3ebd84 libxt_tcp.so`parse_tcp_ports(portstring=0x0000000000000000, ports=0x00000000015e5cd4) + 29 at libxt_tcp.c:46
    frame #3: 0x00007f3b2a3ec191 libxt_tcp.so`tcp_parse(c=50, argv=0x00007f3b2a61ad00, invert=0, flags=0x00007f3b2a61af40, entry=0x00000000014f84f0, match=0x00007f3b2a61aac0) + 289 at libxt_tcp.c:161
    frame #4: 0x00007f3b2c859c46 libxtwrapper.so`wrap_parse(fn=(libxt_tcp.so`tcp_parse at libxt_tcp.c:143), i=50, argv=0x00007f3b2a61ad00, inv=0, flags=0x00007f3b2a61af40, p=0x00000000014f84f0, mptr=0x00007f3b2a61aac0) + 86 at wrapper.c:39
```


The issue is that `optarg` goes from valid -> null after this line (not sure why, I assume that it is not playing nice with uwsgi): <https://github.com/qris/iptables/blob/d225d979534c2accc864504475e927c6b265f9b3/extensions/libxt_tcp.c#L161>.  Assuming that this is not nooby cli program design, my best bet is that it is a result of how iptables handled argument parsing long ago.

Normally one would expect there to be a single option parse somewhere near the programs entry, like [how uwsgi does it](https://github.com/unbit/uwsgi/blob/3dc786f2c8843151612aebdfc479eb41ee7a004f/core/init.c#L256).


To move past the segfault, you can change this:
``` diff
-    parse_tcp_ports(optarg, tcpinfo->dpts);
+    parse_tcp_ports(argv[1], tcpinfo->dpts);
```
However, there seems to be something else afoot here which then causes `XTablesError: DNAT: parameter error -2`.  Still looking in to a full solution.

**Comment by ldx** [comment]
So `optarg` is a global variable in libc - any chance uwsgi changes it?

**Comment by yofreke** [comment]
I ended up just writing a CLI wrapper with a similar API python-iptables.  After debugging / modifying iptables for while, I decided that it was not worth the time.  After looking at the source for a while, my conclusion is that iptables assumes that it is not being bound to, and that it is a standalone cli.  Without a lot of work, it will be difficult to ensure that it plays nice as a library.

I could maybe try to contribute the CLI wrapper i made back to this project if you think it would be worthwhile.
